### ISSUE TITLE
Tolino Epos 2 orientation change - faulty rendering

### ISSUE THREAD

**Issue by mikebluu** [koreader/koreader#10591]
* KOReader version: 2023.05.1
* Device: Tolino Epos 2 (rooted Android 4.4.2)

#### Issue
![PXL_20230617_142623014](https://github.com/koreader/koreader/assets/975883/4112b497-2102-41ab-9620-60f4c98e36c4)

Here again with another bug! Again, since 2023.05, something strange happens when I try changing a document's orientation. It doesn't matter if it's an EPUB or DOCX, they all do the same (even PDFs, rarely but they do it). when I'm changing the orientation from the default (portrait, buttons on the right) to landscape (buttons up), the document is rendered as if the display's width remains the same (as you can see in the [photos](https://photos.app.goo.gl/NdLZSziYzv3xTWUe9), the same when going back from landscape to portrait. If I'm not mistaken, it just renders the document way past the screen's boundaries (e.g. I'm not able to access the bottom menu when my document is rotated like in the second picture)

I don't remember having this problem with earlier versions of KOReader, and I switch orientations frequently. No settings changed also, just upgraded and that's that. If I'm required, I can help you with logs or tests on the device.

#### Steps to reproduce
I'm not that good at explaining, so I made [this](https://youtu.be/Sk1mboAIfKI) video. To get rid of the bug, I just go back home and kill KOReader from the recent apps menu. I didn't find another way to solve it.

Also, when I restart KOReader, for a split second it starts just like before, with half of the screen black. But then it re-renders (I don't know if I used the correct term), and goes back to normal.

**Comment by Frenzie** [comment]
Did you apply any patches or some such?

**Comment by mikebluu** [comment]
Nope, nothing. Just updated to 2023.05.1 and discovered maybe a week ago.

**Comment by Frenzie** [comment]
User patches for KOReader?

**Comment by mikebluu** [comment]
What would those patches look like? As I said before, my KOReader is as vanilla as it can be, just customized to my taste and that's it. If you're talking about plugins, I have all of them activated (by default), except the "Japanese support" one.

**Comment by pazos** [comment]
> since 2023.05 [...] I don't remember having this problem with earlier versions of KOReader

We changed how surfaces work recently.

We were using some workarounds to get screen width and height and now we're using the proper stuff (trust system reported height & width). Sadly this seems to uncover system bugs on some legacy system.

I would suggest to stick with v2023.04 for a while. Lets see if this issue affects other Tolinos or other eink devices.

**Comment by mergen3107** [comment]
On my Onyx Boox Kon-Tiki 2 (Android 10) it goes to this state first described in the OP, then to normal landscape. Previously it was going to notmal landscape directly.

**Comment by mikebluu** [comment]
> On my Onyx Boox Kon-Tiki 2 (Android 10) it goes to this state first described in the OP, then to normal landscape. Previously it was going to notmal landscape directly.

This happens to me sometimes too, for example: this morning, "cold boot" of KOReader, it took some time to open because as before, black area appeared on the bottom part of the screen. I waited maybe 1-2s and it sorted out itself, but as I said earlier, happened even during a normal start so no orientation change from my device's main menu.

**Comment by hugleo** [comment]
This is happening even on my normal Xiaomi phone with Android 14. But this only happens if I disable the progress bar.

![photo_2023-06-18_09-45-39](https://github.com/koreader/koreader/assets/4573505/b8550884-db0a-40c1-89f7-e45963124913)

**Comment by duplode** [comment]
> I would suggest to stick with v2023.04 for a while. Lets see if this issue affects other Tolinos or other eink devices.

It's not an e-ink device, but I had the same issue after installing KOReader 2023.05.1 on a Dell Venue 7 tablet running Android 4.4.4. Downgrading to 2023.04 as suggested has brought landscape back to normal.

**Comment by pazos** [comment]
> It's not an e-ink device, but I had the same issue after installing KOReader 2023.05.1 on a Dell Venue 7 tablet running Android 4.4.4. Downgrading to 2023.04 as suggested has brought landscape back to normal.

Thanks for the feedback. I will take a look in a couple of weeks, after old android deprecation. After all [4.0 - 4.3 are not affected by this issue](https://github.com/koreader/android-luajit-launcher/blob/a21ed93786c272fad4040d94fc2518532ed2bfeb/app/src/main/java/org/koreader/launcher/MainActivity.kt#L481-L484).

Not sure how to fix that, but we can always partially revert 4.4-12 to the old behaviour (as in v2023.04) and apply the current one to android 13.

If somebody finds the same issue on android 13 please comment here.

**Comment by NiLuJe** [comment]
> If somebody finds the same issue on android 13 please comment here.

I did notice something weird during the 5 minutes I actually used it on my phone since the cutout fix, and that was the bottom gesture bar indicator thingy popping up and not going away if I tab out/tab in (on Android 12).

Not sure if related, as I haven't actively used it on Android in years, but I don't recall ever seeing that on the same device (can't recall if the last round of serious testing I did was on 12, though).

TL;DR: Limiting the cutout fix to 13 and up sounds sensible.

**Comment by hugleo** [comment]
@pazos 

Black area also happens on Android 14 like that, disabling progress bar:
https://github.com/koreader/koreader/issues/10591#issuecomment-1596133051

**Comment by stareatmyname** [comment]
This reproduces this issue every single time on one of my android devices (though not the other) in 2023.06, and most of the time in 2023.06.1: set koreader to landscape, lock the screen while inside koreader, then unlock again.

The screen assumes it is now in portrait mode, presumably because my OS (android 9 vendor ROM) has its lock screen locked in portrait, and koreader renders as if in portrait, despite the orientation change (actual orientation in landscape), and the bottom half is rendered offscreen. This can be fixed either by restart, or forcing rotation changes for a redraw, involving two gesture triggers (one to hint a portrait rotation, and another to restore it to landscape) because the buttons aren't rendered onscreen.

From reading the commit history, my best guess is the regression was introduced in https://github.com/koreader/android-luajit-launcher/commit/a9d36a04f4f5a7bddcd46dae5d5cd67031fc7f85, specifically the logic that caches `surfaceWidth` (width also interchangable with height) in `surfaceChanged()` for reuse. This is only a guess, as the issue would not be user-patchable and I don't have the build system for android set up.

My theory is that the `surfaceChanged` callback either isn't ran often enough for all the orientation changes, or there are race conditions in the rendering process where the callback triggers after the rendering process is done calling `getScreenWidth` but `getWidth`'s value changed in the meantime. Whichever the case, and I might have not understood the rationale, but I can't see any good reason to cache `getWidth` as returned by android, which I assume are just getters to private variables, and if so no performance gain over potentially getting an outdated/distorted view of the state of the screen.
